File: library/src/main/java/com/afollestad/aesthetic/AestheticTabLayout.java
package com.afollestad.aesthetic;

import static com.afollestad.aesthetic.Rx.onErrorLogAndRethrow;
import static com.afollestad.aesthetic.TintHelper.createTintedDrawable;
import static com.afollestad.aesthetic.Util.adjustAlpha;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.res.ColorStateList;
import android.support.annotation.ColorInt;
import android.support.design.widget.TabLayout;
import android.util.AttributeSet;
import io.reactivex.Observable;
import io.reactivex.disposables.Disposable;

/** @author Aidan Follestad (afollestad) */
public class AestheticTabLayout extends TabLayout {

  private static final float UNFOCUSED_ALPHA = 0.5f;
  private Disposable indicatorModeSubscription;
  private Disposable bgModeSubscription;
  private Disposable indicatorColorSubscription;
  private Disposable bgColorSubscription;

  public AestheticTabLayout(Context context) {
    super(context);
  }

  public AestheticTabLayout(Context context, AttributeSet attrs) {
    super(context, attrs);
  }

  public AestheticTabLayout(Context context, AttributeSet attrs, int defStyleAttr) {
    super(context, attrs, defStyleAttr);
  }

  private void setIconsColor(int color) {
    final ColorStateList sl =
        new ColorStateList(
            new int[][] {
                new int[] { -android.R.attr.state_selected },
                new int[] { android.R.attr.state_selected }
            },
            new int[] { adjustAlpha(color, UNFOCUSED_ALPHA), color });
    for (int i = 0; i < getTabCount(); i++) {
      final TabLayout.Tab tab = getTabAt(i);
      if (tab != null && tab.getIcon() != null) {
        tab.setIcon(createTintedDrawable(tab.getIcon(), sl));
      }
    }
  }

  @SuppressLint("CheckResult")
  @Override
  public void setBackgroundColor(@ColorInt int color) {
    super.setBackgroundColor(color);
    Aesthetic.get()
        .colorIconTitle(Observable.just(color))
        .take(1)
        .subscribe(
            activeInactiveColors -> {
              setIconsColor(activeInactiveColors.activeColor());
              setTabTextColors(
                  adjustAlpha(activeInactiveColors.inactiveColor(), UNFOCUSED_ALPHA),
                  activeInactiveColors.activeColor());
            });
  }

  @Override
  protected void onAttachedToWindow() {
    super.onAttachedToWindow();

    bgModeSubscription =
        Aesthetic.get()
            .tabLayoutBackgroundMode()
            .compose(Rx.distinctToMainThread())
            .subscribe(
                mode -> {
                  if (bgColorSubscription != null) {
                    bgColorSubscription.dispose();
                  }
                  switch (mode) {
                    case PRIMARY:
                      bgColorSubscription =
                          Aesthetic.get()
                              .colorPrimary()
                              .compose(Rx.distinctToMainThread())
                              .subscribe(
                                  ViewBackgroundAction.create(AestheticTabLayout.this),
                                  onErrorLogAndRethrow());
                      break;
                    case ACCENT:
                      bgColorSubscription =
                          Aesthetic.get()
                              .colorAccent()
                              .compose(Rx.distinctToMainThread())
                              .subscribe(
                                  ViewBackgroundAction.create(AestheticTabLayout.this),
                                  onErrorLogAndRethrow());
                      break;
                    default:
                      throw new IllegalStateException("Unimplemented bg mode: " + mode);
                  }
                },
                onErrorLogAndRethrow());

    indicatorModeSubscription =
        Aesthetic.get()
            .tabLayoutIndicatorMode()
            .compose(Rx.distinctToMainThread())
            .subscribe(
                mode -> {
                  if (indicatorColorSubscription != null) {
                    indicatorColorSubscription.dispose();
                  }
                  switch (mode) {
                    case PRIMARY:
                      indicatorColorSubscription =
                          Aesthetic.get()
                              .colorPrimary()
                              .compose(Rx.distinctToMainThread())
                              .subscribe(
                                  this::setSelectedTabIndicatorColor, onErrorLogAndRethrow());
                      break;
                    case ACCENT:
                      indicatorColorSubscription =
                          Aesthetic.get()
                              .colorAccent()
                              .compose(Rx.distinctToMainThread())
                              .subscribe(
                                  this::setSelectedTabIndicatorColor, onErrorLogAndRethrow());
                      break;
                    default:
                      throw new IllegalStateException("Unimplemented bg mode: " + mode);
                  }
                },
                onErrorLogAndRethrow());
  }

  @Override
  protected void onDetachedFromWindow() {
    if (bgModeSubscription != null) {
      bgModeSubscription.dispose();
    }
    if (indicatorModeSubscription != null) {
      indicatorModeSubscription.dispose();
    }
    if (bgColorSubscription != null) {
      bgColorSubscription.dispose();
    }
    if (indicatorColorSubscription != null) {
      indicatorColorSubscription.dispose();
    }
    super.onDetachedFromWindow();
  }
}

File: library/src/main/java/com/afollestad/aesthetic/AestheticTabLayout.kt
/*
 * Licensed under Apache-2.0
 *
 * Designed and developed by Aidan Follestad (@afollestad)
 */
package com.afollestad.aesthetic

import android.annotation.SuppressLint
import android.content.Context
import android.content.res.ColorStateList
import android.support.annotation.ColorInt
import android.util.AttributeSet
import com.afollestad.aesthetic.TabLayoutBgMode.ACCENT
import com.afollestad.aesthetic.TabLayoutBgMode.PRIMARY
import com.afollestad.aesthetic.actions.ViewBackgroundAction
import com.afollestad.aesthetic.utils.TintHelper.createTintedDrawable
import com.afollestad.aesthetic.utils.adjustAlpha
import com.afollestad.aesthetic.utils.distinctToMainThread
import com.afollestad.aesthetic.utils.onErrorLogAndRethrow
import com.afollestad.aesthetic.utils.one
import com.afollestad.aesthetic.utils.plusAssign
import com.google.android.material.tabs.TabLayout
import io.reactivex.Observable
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.disposables.Disposable
import io.reactivex.functions.Consumer

/** @author Aidan Follestad (afollestad) */
class AestheticTabLayout(
  context: Context,
  attrs: AttributeSet? = null
) : TabLayout(context, attrs) {

  private var topLevelSubs: CompositeDisposable? = null
  private var indicatorColorSubscription: Disposable? = null
  private var bgColorSubscription: Disposable? = null

  companion object {
    const val UNFOCUSED_ALPHA = 0.5f
  }

  private fun setIconsColor(color: Int) {
    val sl = ColorStateList(
        arrayOf(
            intArrayOf(-android.R.attr.state_selected), intArrayOf(android.R.attr.state_selected)
        ),
        intArrayOf(color.adjustAlpha(UNFOCUSED_ALPHA), color)
    )
    for (i in 0 until tabCount) {
      val tab = getTabAt(i)
      if (tab != null && tab.icon != null) {
        tab.icon = createTintedDrawable(tab.icon, sl)
      }
    }
  }

  @SuppressLint("CheckResult")
  override fun setBackgroundColor(@ColorInt color: Int) {
    super.setBackgroundColor(color)
    Aesthetic.get()
        .colorIconTitle(Observable.just(color))
        .one()
        .subscribe { (activeColor, inactiveColor) ->
          setIconsColor(activeColor)
          setTabTextColors(
              inactiveColor.adjustAlpha(UNFOCUSED_ALPHA),
              activeColor
          )
        }
  }

  override fun onAttachedToWindow() {
    super.onAttachedToWindow()
    topLevelSubs = CompositeDisposable()

    topLevelSubs += Aesthetic.get()
        .tabLayoutBackgroundMode()
        .distinctToMainThread()
        .subscribe(
            Consumer {
              bgColorSubscription?.dispose()
              when (it) {
                PRIMARY ->
                  bgColorSubscription = Aesthetic.get()
                      .colorPrimary()
                      .distinctToMainThread()
                      .subscribe(
                          ViewBackgroundAction(
                              this@AestheticTabLayout
                          ),
                          onErrorLogAndRethrow()
                      )
                ACCENT ->
                  bgColorSubscription = Aesthetic.get()
                      .colorAccent()
                      .distinctToMainThread()
                      .subscribe(
                          ViewBackgroundAction(
                              this@AestheticTabLayout
                          ),
                          onErrorLogAndRethrow()
                      )
                else -> throw IllegalStateException("Unimplemented bg mode: $it")
              }
            },
            onErrorLogAndRethrow()
        )

    topLevelSubs += Aesthetic.get()
        .tabLayoutIndicatorMode()
        .distinctToMainThread()
        .subscribe(
            Consumer {
              indicatorColorSubscription?.dispose()
              when (it) {
                PRIMARY ->
                  indicatorColorSubscription = Aesthetic.get()
                      .colorPrimary()
                      .distinctToMainThread()
                      .subscribe(
                          Consumer { color -> setSelectedTabIndicatorColor(color) },
                          onErrorLogAndRethrow()
                      )
                ACCENT ->
                  indicatorColorSubscription = Aesthetic.get()
                      .colorAccent()
                      .distinctToMainThread()
                      .subscribe(
                          Consumer { color -> setSelectedTabIndicatorColor(color) },
                          onErrorLogAndRethrow()
                      )
                else -> throw IllegalStateException("Unimplemented bg mode: $it")
              }
            },
            onErrorLogAndRethrow()
        )
  }

  override fun onDetachedFromWindow() {
    topLevelSubs?.dispose()
    bgColorSubscription?.dispose()
    indicatorColorSubscription?.dispose()
    super.onDetachedFromWindow()
  }
}
